src/annotator/anchoring/types.js

```diff
@@ -0,0 +1,245 @@
+    };
+  }
+}
+
+/**
+ * Converts between TextPositionSelector selectors and Range objects.
+ */
+export class TextPositionAnchor {
+  /**
+   * @param {Node} root
+   * @param {number} start
+   * @param {number} end
+   */
+  constructor(root, start, end) {
+    this.root = root;
+    this.start = start;
+    this.end = end;
+  }
+
+  /**
+   * @param {Node} root
+   * @param {Range} range
+   */
+  static fromRange(root, range) {
+    const selector = posFromRange(root, range);
+    return TextPositionAnchor.fromSelector(root, selector);
+  }
+  /**
+   * @param {Node} root
+   * @param {TextPositionSelector} selector
+   */
+  static fromSelector(root, selector) {
+    return new TextPositionAnchor(root, selector.start, selector.end);
+  }
+
+  toSelector() {
+    /** @type {TextPositionSelector} */
+    return {
+      type: 'TextPositionSelector',
+      start: this.start,
+      end: this.end,
+    };
+  }
+
+  toRange() {
+    return posToRange(this.root, { start: this.start, end: this.end });
+  }
+}
+
+/**
+ * Converts between TextQuoteSelector selectors and Range objects.
+ */
+export class TextQuoteAnchor {
+  /**
+   * @param {Node|TextContentNode} root - A root element from which to anchor.
+   * @param {string} exact
+   * @param {TextQuoteSelector} context
+   */
+  constructor(root, exact, context) {
+    if (!context) {
+      context = /** @type {TextQuoteSelector} */ ({});
+    }
+    this.root = root;
+    this.exact = exact;
+    this.context = context;
+  }
+  /**
+   * @param {Node} root
+   * @param {Range} range
+   * @param {SelectorOptions} options
+   */
+  static fromRange(root, range, options) {
+    const selector = quoteFromRange(root, range, options);
+    return TextQuoteAnchor.fromSelector(root, selector);
+  }
+
+  /**
+   * @param {Node|TextContentNode} root
+   * @param {TextQuoteSelector} selector
+   */
+  static fromSelector(root, selector) {
+    const { prefix, suffix } = selector;
+    return new TextQuoteAnchor(
+      root,
+      selector.exact,
+      /** @type {TextQuoteSelector} */ ({ prefix, suffix })
+    );
+  }
+
+  toSelector() {
+    /** @type {TextQuoteSelector} */
+    return {
+      type: 'TextQuoteSelector',
+      exact: this.exact,
+      prefix: this.context.prefix,
+      suffix: this.context.suffix,
+    };
+  }
+
+  /**
+   * @param {SelectorOptions} options
+   */
+  toRange(options) {
+    if (!options) {
+      options = {};
+    }
+    const range = quoteToRange(this.root, this.toSelector(), options);
+    if (range === null) {
+      throw new Error('Quote not found');
+    }
+    return range;
+  }
+
+  /**
+   * @param {SelectorOptions} [options]
+   */
+  toPositionAnchor(options) {
+    if (!options) {
+      options = {};
+    }
+    const anchor = toTextPosition(this.root, this.toSelector(), options);
+    if (anchor === null) {
+      throw new Error('Quote not found');
+    }
+    return new TextPositionAnchor(
+      /** @type {Node} */ (this.root),
```

Comment:
This cast isn't safe if `TextQuoteAnchor` was initialized with a `TextContentNode` that is not a `Node`. It looks like `TextPositionAnchor`'s constructor can be changed to accept a `TextContentNode` instead.